### DISCUSSION TITLE
After deleting a dag from the dags directory, it is still displayed in the UI

### DISCUSSION THREAD

**Discussion by tatyana12345** [apache/airflow#40331] [Q&A]
### Apache Airflow version

Other Airflow 2 version (please specify below)

### If "Other Airflow 2 version" selected, which one?

2.8.3

### What happened?

Dag has been removed from the airflow dags directory /opt/airflow/dags. Dag is not displayed when running the airflow dags list command. But it is still available in the Airflow UI, and it is also available in the Airflow postgres database.

### What you think should happen instead?

It is expected that dag will be automatically removed from the web interface after deleting dag from the dags directory.

### How to reproduce

Remove dag from the /opt/airflow/dags directory and it will still be available in the web interface even after a few days

### Operating System

Linux

### Deployment

Official Apache Airflow Helm Chart

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

**Comment by boring-cyborg** [comment]
Thanks for opening your first issue here! Be sure to follow the issue template! If you are willing to raise PR to address this issue please do so, no need to wait for approval.

**Comment by tirkarthi** [comment]
Deleting the Airflow dag file doesn't delete the db entries. You might want to try below command to delete the related entries. Please also check the below discussion that has different steps in deleting the dag completely.

https://github.com/apache/airflow/discussions/21864

```
airflow dags delete -h 
Usage: airflow dags delete [-h] [-v] [-y] dag_id

Delete all DB records related to the specified DAG

Positional Arguments:
  dag_id         The id of the dag

Options:
  -h, --help     show this help message and exit
  -v, --verbose  Make logging output more verbose
  -y, --yes      Do not prompt to confirm. Use with care!
```

**Reply by tatyana12345** [reply]
Thanks for your reply! I understand that I can delete a dag via the CLI or even from the database, but it seems inconvenient to delete the same dag from two places.
Do you have any information on why deleting the Airflow DAG file does not delete the database entries? This worked in previous versions, but no longer works in the current one.

**Reply by tirkarthi** [reply]
I am not sure such feature was present. Deleting dags in dags folder stops dag processor from picking it up and after a certain point depending on configurations the value is deleted in serialized_dag table with dag deactivated. But the metadata linked to dag still had to be manually deleted. This could be inconvenient and confusing but that's how it worked in the past and now. Adding relevant part from the discussion.

> This all means that if you want to actually delete a DAG and its all historical metadata, you need to do
it in three steps:
> pause the DAG
> delete the historical metadata from the database, via UI or API
> delete the DAG file from the DAGS_FOLDER and wait until it becomes inactive

**Reply by falukelo** [reply]
I got this issue, I am try this suggestion, but the deleted DAGs keep reappearing.

**Reply by gaelxcowi** [reply]
@tirkarthi I am not sure if I am miss understading or the discussion that you ahve pointed to https://github.com/apache/airflow/discussions/21864 contradicts what you are saying. 

I am experiencing the same behaviour as the OP, upon deleting the dag file from the dag directory, I expect a few things to happen: 

1. Dag is marked as  "deactivated" in the db 
2. Dag is NOT displayed in the main view of the dags (some references of the dag might still be present when browsing dag runs/task instances
3. Dag metadata is still kept in the db 
4. When adding back the dag file, the history is presereved. 

I am getting this from the discussion you have linked, specifically: 

_Dag can be deactivated (do not confuse it with Active tag in the UI by removing them from the
DAGS_FOLDER. When scheduler parses the DAGS_FOLDER and misses the DAG that it had seen
before and stored in the database it will set is as **deactivated**. The metadata and history of the
DAG` is kept for deactivated DAGs and when the DAG is re-added to the DAGS_FOLDER it will be again
activated and history will be visible. You cannot activate/deactivate DAG via UI or API, this
can only be done by removing files from the ``DAGS_FOLDER``. Once again - no data for historical runs of the
DAG are lost when it is deactivated by the scheduler. Note that the ``Active`` tab in Airflow UI
refers to DAGs that are both ``Activated`` and ``Not paused`` so this might initially be a
little confusing._

_You **can't see the deactivated DAGs in the UI** - you can sometimes see the historical runs, but when you try to see the information about those you will see the error that the DAG is missing._

What I am experiencing right now, is after removing the dag from the DAGS_FOLDER, nothing really happens, the dag is still visible, accessible and interactable (can run, pause, clear, etc.) in the UI (which suggests that it was never actually removed from the serialized table) tho if attempting to run a task the executor will error out because the dag is missing or cannot be parsed. 

The airflow dags list only shows the dags that have a file in the DAGS_FOLDER. 

The scheduler errors out with "File not found" error of the removed dag file throwin by the DagFileProcessor process. 



So, again not sure if I am missunderstanding, but as the OP mentioned, the behaviour described in the conversation that you have linked to me contradicts with what you are saying that the dag HAS TO be deleted completle from the metadata db before it won't be shown in the UI. 

Also this WAS NOT the behaviour in the past, in the past it would follow the description that you have linked.

Airflow 2.9.3

**Reply by cjj1120** [reply]
@gaelxcowi having the exact same issue with airflow 2.9.3

**Comment by shahar1** [comment]
Please try @tirkarthi 's suggestion.
I'm converting the issue into a discussion, as it seems like an expected behavior.

**Reply by gaelxcowi** [reply]
@shahar1 how is this the expected behaviour - if literally in the last comment I explain how it is the opposite of the documented behaviour with no reply?

**Reply by shahar1** [reply]
> @shahar1 how is this the expected behaviour - if literally in the last comment I explain how it is the opposite of the documented behaviour with no reply?

Your detailed comment did not exist at the time - so I didn't have much to do.
Also, I've tried reproduce the issue according to your comment using local `breeze` deployment (default configuration) - but it seems to work as expected: after deleting the DAG from the DAGs folder, it exists in the UI for a couple more seconds and then removed from there.
As there are 4 reports of this problem in this discussion, I [reopened](https://github.com/apache/airflow/issues/42542) it back as an issue. However, we'll need your help providing the exact Airflow configuration, deployment environment, Airflow version, and steps to reproduce the issue (preferrably with accompanying timestamps and images from the UI). We will also be happy if you could make some effort contributing a PR to solve it, I'll be more than happy to review it.

**Reply by gaelxcowi** [reply]
Okay, my bad - did not realize that my comment was after you have closed. Will try to gather all the evidence as soon as possible

**Comment by shahar1** [comment]
Closing as reopened back as an [issue](https://github.com/apache/airflow/issues/42542).
